### PR TITLE
Propagate shape information for View support - V2

### PR DESCRIPTION
**Pull request by rdspring1**
Add shape information to outputs via `PropagateShapesOnGraph`
View and Reshape need the original size of the tensor to find which axes to merge, split, and broadcast.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 1]
```diff
@@ -153,155 +172,124 @@ class NaiveTypePropagator {
...
+        // TODO: double check with type promotion
+        if (output_mask[1]) {
+          if (auto weight_type = getInputTensorType(node, 3, true)) {
+            node->output(1)->setType(weight_type);
+          }
+        }
 
+        // TODO: Use shape information from weight tensor
+        if (output_mask[2]) {
+          auto bias_type = TensorType::create(
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp:251]: I thought we mentioned about `at::toAccumulateType` to use here insterad? (same as with `weight_type`), even though it's a hard to maintain thing

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 2]
```diff
@@ -153,155 +172,124 @@ class NaiveTypePropagator {
...
-
-        node->output(1)->setType(mean_rstd_type);
-        node->output(2)->setType(mean_rstd_type);
-        // TODO: not that it matters, but mark the right type here;
-        // node->output(3)->setType(out_type->withScalarType());
-        node->output(3)->setType(out_type);
-        node->output(4)->setType(IntType::get());
+        // Assume type promotion would have consistent dtype
+        // between batch stats and running stats
+        if (auto mean_type = getInputTensorType(node, 3, true)) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp]: Same here, also running stats output is not necessarily bundled with weight/bias. So the condition statement here is not right.

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 3]
```diff
@@ -153,155 +172,124 @@ class NaiveTypePropagator {
...
 
-        break;
-      }
-      case aten::native_batch_norm: {
-        auto out_type = node->input(0)->type()->cast<TensorType>();
-        TORCH_CHECK(
-            hasTypeAndDevice(out_type),
-            "Type and device propagation has failed, or was not provided enough information.");
-        node->output(0)->setType(out_type);
+        if (auto var_type = getInputTensorType(node, 4, true)) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp]: ditto

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 4]
```diff
@@ -153,155 +172,124 @@ class NaiveTypePropagator {
...
-          auto in_type = node->input(1)->type()->cast<TensorType>();
-          node->output(0)->setType(in_type);
-        }
+      case aten::native_batch_norm: {
+        node->output(0)->setType(getInputTensorType(node, 0));
 
-        if (output_mask[1]) {
-          auto weight_type = node->input(2)->type()->cast<TensorType>();
-          node->output(1)->setType(weight_type);
+        if (auto mean_type = getInputTensorType(node, 3, true)) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp]: ditto

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 5]
```diff
@@ -153,155 +172,124 @@ class NaiveTypePropagator {
...
 
-        if (output_mask[2]) {
-          auto weight_type = node->input(2)->type()->cast<TensorType>();
-          auto bias_type = TensorType::create(
-              *weight_type->scalarType(),
-              *weight_type->device(),
-              *weight_type->dim(),
-              output_mask[2]);
-          node->output(2)->setType(bias_type);
+        if (auto var_type = getInputTensorType(node, 4, true)) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp]: ditto

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 6]
```diff
@@ -153,155 +172,124 @@ class NaiveTypePropagator {
...
-        node->output()->setType(out_type);
+        node->output(0)->setType(getInputTensorType(node, 0));
         break;
       }
       case aten::native_layer_norm: {
-        auto out_type = node->input(0)->type()->cast<TensorType>();
-        TORCH_CHECK(
-            hasTypeAndDevice(out_type),
-            "Type and device propagation has failed, or was not provided enough information.");
+        auto out_type = getInputTensorType(node, 0);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp:300]: This is not in your code. But can we also update the `mean_rstd_type` :pray:

**Review by jjsjann123 — COMMENTED** [review]
LGTM with minor comments.

File: torch/csrc/jit/codegen/cuda/type_inference.cpp (hunk 1 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/type_inference.cpp:251]: You mentioned that you were "already patching that on a unrelated issue", so I wasn't sure who was going to add the accumulate type rules.

Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/type_inference.cpp:251]: I'll go ahead and add them.

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 7]
```diff
@@ -15,10 +15,64 @@ namespace cuda {
 
 namespace {
 
+at::ScalarType toAccumulateType(const TensorTypePtr& op) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp:19]: Any reason that we are not using `at::toAccumulateType` from `aten/src/ATen/AccumulateType.h`?

**Review by jjsjann123 — APPROVED** [review]
LGTM, trivial comment on potential duplicated code.

File: torch/csrc/jit/codegen/cuda/type_inference.cpp (hunk 7 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/type_inference.cpp:19]: No good reason I guess.

**Pull request merged**